### PR TITLE
Add an sql function for carving paths

### PR DESCRIPTION
**Pull request by jedi22**

### CODE REVIEW

**Comment by osqueryer** [comment]
:-1: The commit 4d806ed6d84266baf33e0dafcfcad4d5da6e4c93 (Job results: [1679](https://jenkins.osquery.io/job/osqueryPullRequestWindows/1679/)) failed one or more tests (**Windows**).

**Comment by osqueryer** [comment]
:-1: The commit 4d806ed6d84266baf33e0dafcfcad4d5da6e4c93 (Job results: [1680](https://jenkins.osquery.io/job/osqueryPullRequestWindows/1680/)) failed one or more tests (**Windows**).

File: osquery/core/system.cpp [hunk 1]
```diff
@@ -471,4 +477,29 @@ DropPrivileges::~DropPrivileges() {
...
+  if (paths.size() > 1) {
+    tree.put("path", boost::algorithm::join(paths, ","));
+  } else {
+    tree.put("path", *(paths.begin()));
+  }
+  std::ostringstream os;
+  pt::write_json(os, tree, false);
+  auto s = setDatabaseValue(kCarveDbDomain, kCarverDBPrefix + guid, os.str());
+  if (!s.ok()) {
+    return s;
```
Comment (poppyseedplehzr) [inline osquery/core/system.cpp]: It doesn't look like you leverage this return value anywhere. Perhaps let's just make this function a `void` and `VLOG(1)` if setting the database value fails?

Comment (jedi22) [inline osquery/core/system.cpp]: I don't personally use it my code path but I don't want to preclude people from having access to it in the future.

**Comment by osqueryer** [comment]
:no_good: The commit 1d051633a57a7fa5b29fae2481c6da94b2e2ab1c (Job results: [1622](https://jenkins.osquery.io/job/osqueryPullRequestAudit/1622/)) failed the code audit or documentation test.

**Comment by osqueryer** [comment]
:-1: The commit 5d9d78c3aca233365b969f5371f14b41d14d66ef (Job results: [1682](https://jenkins.osquery.io/job/osqueryPullRequestWindows/1682/)) failed one or more tests (**Windows**).

**Comment by osqueryer** [comment]
:-1: The commit 5d9d78c3aca233365b969f5371f14b41d14d66ef (Job results: [1683](https://jenkins.osquery.io/job/osqueryPullRequestWindows/1683/)) failed one or more tests (**Windows**).

**Comment by jedi22** [comment]
@poppyseedplehzr Do you know off hand why only Windows is throwing this error?

**Comment by osqueryer** [comment]
:no_good: The commit 48f6660ad57274b5d15575c0f18a7f48b9403ce4 (Job results: [1649](https://jenkins.osquery.io/job/osqueryPullRequestAudit/1649/)) failed the code audit or documentation test.

**Comment by osqueryer** [comment]
:-1: The commit e701bf8e155a694e82e5ac53fbcddd4290172e11 (Job results: [1712](https://jenkins.osquery.io/job/osqueryPullRequestWindows/1712/)) failed one or more tests (**Windows**).

**Comment by jedi22** [comment]
ok to test

**Comment by osqueryer** [comment]
:no_good: The commit f92424cf9830b521ead1daeb013976a14869ad4e (Job results: [1652](https://jenkins.osquery.io/job/osqueryPullRequestAudit/1652/)) failed the code audit or documentation test.

**Comment by osqueryer** [comment]
:-1: The commit f92424cf9830b521ead1daeb013976a14869ad4e (Job results: [4839](https://jenkins.osquery.io/job/osqueryPullRequestLinux/4839/)) failed one or more tests (**Linux**).

File: osquery/carver/carver.cpp [hunk 2]
```diff
@@ -318,4 +313,29 @@ Status Carver::postCarve(const boost::filesystem::path& path) {
...
+  tree.put("time", getUnixTime());
+  tree.put("status", "STARTING");
+  tree.put("sha256", "");
+  tree.put("size", -1);
+  if (paths.size() > 1) {
+    tree.put("path", boost::algorithm::join(paths, ","));
+  } else {
+    tree.put("path", *(paths.begin()));
+  }
+  std::ostringstream os;
```
Comment (theopolis) [inline osquery/carver/carver.cpp]: This code was moved and the spaces were left out, IMO a double newline goes a long way for readability.

File: osquery/carver/carver.cpp [hunk 3]
```diff
@@ -66,17 +68,10 @@ CLI_FLAG(bool,
...
-
-/// Prefix used for the temp FS where carved files are stored
-const std::string kCarvePathPrefix = "osquery_carve_";
-
-/// Prefix applied to the file carve tar archive.
-const std::string kCarveNamePrefix = "carve_";
-
-/// Database prefix used to directly access and manipulate our carver entries
-const std::string kCarverDBPrefix = "carves.";
+CLI_FLAG(bool,
```
Comment (theopolis) [inline osquery/carver/carver.cpp]: Why?

File: osquery/carver/carver.h [hunk 4]
```diff
@@ -15,6 +15,18 @@
 
 namespace osquery {
 
+/// Database domain where we store carve table entries
+const std::string kCarveDbDomain = "carves";
```
Comment (theopolis) [inline osquery/carver/carver.h]: Do not define in headers please.

File: osquery/core/system.cpp [hunk 5]
```diff
@@ -39,6 +39,7 @@
 
 #include <osquery/core.h>
 #include <osquery/database.h>
+#include <osquery/dispatcher.h>
```
Comment (theopolis) [inline osquery/core/system.cpp]: Why all of these new includes?

File: osquery/core/system.cpp [hunk 6]
```diff
@@ -48,11 +49,14 @@
...
 #include "osquery/core/windows/wmi.h"
 #endif
+#include "osquery/carver/carver.h"
 #include "osquery/core/conversions.h"
+#include "osquery/core/json.h"
 #include "osquery/core/process.h"
 #include "osquery/core/utils.h"
 
 namespace fs = boost::filesystem;
+namespace pt = boost::property_tree;
```
Comment (theopolis) [inline osquery/core/system.cpp]: Why?

File: osquery/sql/sqlite_operations.cpp [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+ *  Copyright (c) 2014-present, Facebook, Inc.
+ *  All rights reserved.
+ *
+ *  This source code is licensed under the BSD-style license found in the
+ *  LICENSE file in the root directory of this source tree. An additional grant
+ *  of patent rights can be found in the PATENTS file in the same directory.
+ *
+ */
+
+#include <functional>
```
Comment (theopolis) [inline osquery/sql/sqlite_operations.cpp]: Do you need this?

File: osquery/sql/sqlite_operations.cpp [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+#include <set>
+#include <string>
+
+#include <osquery/flags.h>
+#include <osquery/logger.h>
+
+#include "osquery/carver/carver.h"
+#include "osquery/core/conversions.h"
+#include "osquery/tables/system/hash.h"
```
Comment (theopolis) [inline osquery/sql/sqlite_operations.cpp]: Do you need this?

File: osquery/carver/carver.cpp (hunk 3 above)
Comment (jedi22) [inline osquery/carver/carver.cpp]: So we can disable just the function while still leaving the table.

File: osquery/carver/carver.h (hunk 4 above)
Comment (jedi22) [inline osquery/carver/carver.h]: What if multiple files need to access these constants?

**Pull request merged**